File: apple/HavenApp/ActivityStore.swift
import Foundation
import SwiftUI

/// The in-app activity list (the bell): who reacted / commented / voted / posted / messaged,
/// merged from two sources —
///   • the engine's `activity(sinceMs:nowMs:)` reduction (one pass across every circle, own
///     actions excluded), pulled OFF-MAIN on the engine actor after ingest bursts and on foreground;
///   • app-layer moments that never enter a circle's event log — connection requests,
///     added-to-a-circle, device linking — appended at the same sites that post their banners.
///
/// Rows persist as a small JSON file in Application Support (cap ~500) so the list survives a
/// relaunch. `seenAt` is the read watermark: unread = rows newer than it, and it syncs across
/// the user's own devices via SelfSync (`setting:activitySeenAt`, merged MAX — the exact
/// dmLastRead pattern), so opening the bell on one device clears it everywhere.
@MainActor
final class ActivityStore: ObservableObject {
    static let shared = ActivityStore()

    struct Entry: Codable, Identifiable, Equatable {
        /// Event id (engine rows) or the banner dedupe key (app-layer rows) — the dedupe axis.
        let id: String
        /// `react|comment|vote|post|story|dm` (engine) or `connect|circle|device` (app-layer).
        let kind: String
        let circleId: String
        let actorHex: String
        /// 8-char author prefix for `ContactsStore.name(forNodePrefix:)`.
        let actorShort: String
        /// The parent post/comment/poll id where applicable (reactions, comments, votes).
        let targetId: String?
        let snippet: String
        let emoji: String?
        /// Event time, unix ms.
        let at: UInt64
    }

    @Published private(set) var entries: [Entry] = []
    /// Rows newer than the seen watermark — the bell badge.
    @Published private(set) var unread = 0

    private var ids = Set<String>()
    private var pullInFlight = false
    private var savePending = false
    private static let cap = 500

    private let seenAtKey = "haven.activity.seenAt"
    /// The read watermark (unix ms). Monotonic — see `applySyncedSeenAt`.
    private(set) var seenAtMs: UInt64

    private let pulledThroughKey = "haven.activity.pulledThroughMs"
    /// The newest EVENT TIME the engine's `activity()` has actually handed us — the ONLY thing
    /// allowed to narrow the next pull's window.
    ///
    /// This used to be `entries.first?.at`: the newest row in the list, whatever its source. But
    /// the list is a MERGE, and its other source stamps `Date()`. One `note(…)` row — "someone
    /// wants to connect", "you were added to a circle", "a device was linked" — therefore set the
    /// engine's watermark to *now*, and every pull after it asked the engine for the last hour
    /// only. Anything older that the list had not already ingested was skipped, and skipped again
    /// on the next pull, because that app-layer row stayed the newest one: a permanent hole, not a
    /// slow refresh. `noteFeedItem` did the same with a real event time — `runCircleSideEffects`
    /// mirrors the newest inbound item and *then* pulls, so a phone catching up after a day
    /// offline noted one fresh message and clamped the pull that was supposed to fetch the
    /// backlog behind it.
    ///
    /// Persisted, so a relaunch keeps the incremental window. Absent (or 0) means "ask for
    /// everything" — which is what a fresh install does anyway, and what repairs a list this bug
    /// already punched a hole in.
    private var pulledThroughMs: UInt64

    private var fileURL: URL {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("haven-activity.json")
    }

    private init() {
        seenAtMs = UInt64(max(0, UserDefaults.standard.double(forKey: seenAtKey)))
        pulledThroughMs = UInt64(max(0, UserDefaults.standard.double(forKey: pulledThroughKey)))
        if let data = try? Data(contentsOf: fileURL),
           let rows = try? JSONDecoder().decode([Entry].self, from: data) {
            entries = rows
            ids = Set(rows.map(\.id))
        }
        recomputeUnread()
    }

    // MARK: - Intake

    /// Engine rows. The FFI runs on the engine actor (it takes the engine lock); the merge
    /// publishes back on the main actor. Coalesced — an ingest burst pulls once.
    func pull(engine: Engine) {
        guard !pullInFlight else { return }
        pullInFlight = true
        // Overlap what the engine has already given us by an hour so an event that raced the
        // previous pull isn't skipped; `ingest` dedupes by event id, so overlap costs nothing.
        // The watermark is `pulledThroughMs` and NOT the newest row in the list — see its comment.
        let since = pulledThroughMs > 3_600_000 ? pulledThroughMs - 3_600_000 : 0
        let nowMs = UInt64(Date().timeIntervalSince1970 * 1000)
        Task { @MainActor [weak self] in
            let rows = await engine.run { $0.activity(sinceMs: since, nowMs: nowMs) }
            guard let self else { return }
            self.pullInFlight = false
            self.ingest(rows)
        }
    }

    /// Merge one completed engine pull. This is the only place the pull watermark moves, and it
    /// moves to an EVENT time the engine reported — never to the wall clock, so a backfill that
    /// delivers older events later still lands.
    private func ingest(_ rows: [ActivityItemFfi]) {
        if let newest = rows.map(\.createdAt).max(), newest > pulledThroughMs {
            pulledThroughMs = newest
            scheduleSave()   // the watermark rides the SAME write as the rows it accounts for
        }
        var added = false
        for r in rows where !ids.contains(r.id) {
            ids.insert(r.id)
            entries.append(Entry(id: r.id, kind: r.kind, circleId: r.circleId,
                                 actorHex: r.actorHex, actorShort: r.actorShort,
                                 targetId: r.targetId, snippet: r.snippet, emoji: r.emoji,
                                 at: r.createdAt))
            added = true
        }
        if added { finalizeMutation() }
    }

    /// App-layer rows (connection request, added-to-circle, device linked) — appended at the same
    /// sites that post their banners, so the list is complete even when banners are suppressed.
    func note(id: String, kind: String, circleId: String = "", actorHex: String = "",
              actorShort: String = "", snippet: String) {
        guard !ids.contains(id) else { return }
        ids.insert(id)
        entries.append(Entry(id: id, kind: kind, circleId: circleId, actorHex: actorHex,
                             actorShort: actorShort, targetId: nil, snippet: snippet,
                             emoji: nil, at: UInt64(Date().timeIntervalSince1970 * 1000)))
        finalizeMutation()
    }

    /// A feed item the banner path just surfaced (`applyNotifyFromItems`) — mirror it into the
    /// list ahead of the next engine pull. Same id as the engine's row, so the pull dedupes.
    func noteFeedItem(_ item: FeedItemFfi, circleId: String) {
        guard !ids.contains(item.id) else { return }
        ids.insert(item.id)
        let kind = item.story ? "story" : (circleId.hasPrefix("dm:") ? "dm" : "post")
        entries.append(Entry(id: item.id, kind: kind, circleId: circleId, actorHex: "",
                             actorShort: item.authorShort, targetId: nil,
                             snippet: String(item.body.prefix(120)), emoji: nil,
                             at: item.createdAt))
        finalizeMutation()
    }

    private func finalizeMutation() {
        entries.sort { $0.at > $1.at }
        if entries.count > Self.cap {
            for e in entries.suffix(entries.count - Self.cap) { ids.remove(e.id) }
            entries.removeLast(entries.count - Self.cap)
        }
        recomputeUnread()
        scheduleSave()
    }

    // MARK: - Seen watermark

    /// Opening the activity list clears the bell — here, and (via SelfSync) on every device.
    func markAllSeen() {
        let now = UInt64(Date().timeIntervalSince1970 * 1000)
        guard now > seenAtMs else { return }
        seenAtMs = now
        UserDefaults.standard.set(Double(seenAtMs), forKey: seenAtKey)
        recomputeUnread()
        // A LOCAL clear (never `applySyncedSeenAt`) drops the bell badge on my other devices in
        // seconds via a debounced forced self-sync pass.
        FeedStore.shared.nudgeSelfSyncSoon()
    }

    /// SelfSync apply (`setting:activitySeenAt`): merged MAX — reading on one device clears the
    /// bell everywhere, and no device can un-read another (a fresh device's 0 changes nothing).
    func applySyncedSeenAt(_ ms: UInt64) {
        guard ms > seenAtMs else { return }
        seenAtMs = ms
        UserDefaults.standard.set(Double(seenAtMs), forKey: seenAtKey)
        recomputeUnread()
    }

    private func recomputeUnread() {
        let n = entries.reduce(0) { $0 + ($1.at > seenAtMs ? 1 : 0) }
        if n != unread { unread = n }
    }

    // MARK: - Persistence (debounced, off-main write)

    private func scheduleSave() {
        guard !savePending else { return }
        savePending = true
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
            guard let self else { return }
            self.savePending = false
            // Watermark and rows are written together on purpose. Persisting `pulledThroughMs` the
            // moment a pull returned would narrow the next launch's window on the strength of rows
            // that a kill in the next two seconds never wrote — a hole that no later pull would
            // ever ask for again. Written here they advance together, and a kill in between simply
            // re-pulls the same window.
            UserDefaults.standard.set(Double(self.pulledThroughMs), forKey: self.pulledThroughKey)
            guard let data = try? JSONEncoder().encode(self.entries) else { return }
            let url = self.fileURL
            DispatchQueue.global(qos: .utility).async {
                try? data.write(to: url, options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
            }
        }
    }
}

File: android/app/src/main/java/com/blaineam/haven/core/ActivityStore.kt
package com.blaineam.haven.core

import android.content.Context
import androidx.compose.runtime.mutableIntStateOf
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONObject
import uniffi.haven_ffi.HavenSocial

/**
 * The in-app activity list — "who did what that concerns me", across every circle (parity with the
 * core `social.activity()` reducer the other platforms share).
 *
 * Two row sources merge here:
 *  • ENGINE rows: `social.activity()` reduces every circle's decrypted events (reactions/comments/
 *    votes on MY events, others' posts, stories, DMs) — recomputed off-main on a debounced poke
 *    from HavenNet's ingest paths, never persisted (the event log is the store).
 *  • APP rows: things the engine can't see — a new connection accepted, being added to a circle —
 *    recorded by HavenNet where it detects them, persisted as JSON in SharedPreferences.
 *
 * `seenAt` is the read watermark: rows newer than it badge the bell. It self-syncs across the
 * user's devices via `setting:activitySeenAt` (per-key MAX merge — reading on one device clears
 * the badge on the others, and no device can un-read another).
 */
object ActivityStore {
    private const val PREFS = "haven.activity"
    private const val KEY_SEEN_AT = "seenAt"
    private const val KEY_APP_ROWS = "appRows"
    private const val ENGINE_WINDOW_MS = 30L * 24 * 3600 * 1000   // a month of history is plenty

    /** One rendered row. [kind] is the core's ("react"|"comment"|"vote"|"post"|"story"|"dm") plus
     *  the app-layer "connect" and "circle". [targetId] = the parent post for reactions/comments. */
    data class Row(
        val id: String,
        val kind: String,
        val circleId: String,
        val actorHex: String,
        val actorShort: String,
        val targetId: String?,
        val snippet: String,
        val createdAt: Long,
        val emoji: String?,
    )

    private lateinit var appContext: Context
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val prefs get() = appContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    /** Bumped whenever [rows] or [seenAt] change so the bell badge + list recompose. */
    val version = mutableIntStateOf(0)

    /** Merged rows, newest-first. Read from the UI; refreshed off-main by [poke]. */
    @Volatile var rows: List<Row> = emptyList()
        private set

    @Volatile private var seenAtMs: Long = 0L
    @Volatile private var refreshPending = false

    fun init(context: Context) {
        if (this::appContext.isInitialized) return
        appContext = context.applicationContext
        seenAtMs = prefs.getLong(KEY_SEEN_AT, 0L)
        rows = loadAppRows()
    }

    private val ready get() = this::appContext.isInitialized

    // ---- Read watermark ------------------------------------------------------------------------

    fun seenAt(): Long = seenAtMs

    fun unseenCount(): Int = rows.count { it.createdAt > seenAtMs }

    /** The user looked at the list — everything currently shown is read. Self-syncs (MAX merge). */
    fun markSeen() {
        val now = System.currentTimeMillis()
        if (now <= seenAtMs) return
        seenAtMs = now
        if (ready) prefs.edit().putLong(KEY_SEEN_AT, now).apply()
        bump()
        // LOCAL read of the bell — nudge a debounced forced self-sync so the badge clears on my
        // other devices in seconds (applySyncedSeenAt, the remote path, never lands here).
        HavenNet.selfSyncNudge()
    }

    /** A sibling device's watermark arrived via self-sync — per-key MAX merge (monotonic). */
    fun applySyncedSeenAt(ms: Long) {
        if (ms <= seenAtMs) return
        seenAtMs = ms
        if (ready) prefs.edit().putLong(KEY_SEEN_AT, ms).apply()
        bump()
    }

    // ---- App-layer rows (connections / circle adds) --------------------------------------------

    /** A new connection was accepted — the engine has no event for it, so record it here. */
    fun noteConnection(idHex: String, name: String) {
        addAppRow(Row(
            id = "connect:$idHex",
            kind = "connect",
            circleId = "",
            actorHex = idHex,
            actorShort = idHex.take(8),
            targetId = null,
            snippet = name,
            createdAt = System.currentTimeMillis(),
            emoji = null,
        ))
    }

    /** This device learned it was added to a (new) circle. */
    fun noteCircleAdd(circleId: String, circleName: String) {
        addAppRow(Row(
            id = "circle:$circleId",
            kind = "circle",
            circleId = circleId,
            actorHex = "",
            actorShort = "",
            targetId = null,
            snippet = circleName,
            createdAt = System.currentTimeMillis(),
            emoji = null,
        ))
    }

    private fun addAppRow(row: Row) {
        if (!ready) return
        val existing = loadAppRows()
        if (existing.any { it.id == row.id }) return   // deduped on the stable id
        val next = (listOf(row) + existing).take(200)
        saveAppRows(next)
        poke(null)   // remerge (engine rows unchanged; social may be absent — app rows still land)
        bump()
    }

    private fun loadAppRows(): List<Row> {
        if (!ready) return emptyList()
        val raw = prefs.getString(KEY_APP_ROWS, null) ?: return emptyList()
        return runCatching {
            val arr = JSONArray(raw)
            (0 until arr.length()).mapNotNull { i ->
                val o = arr.optJSONObject(i) ?: return@mapNotNull null
                Row(
                    id = o.optString("id"),
                    kind = o.optString("kind"),
                    circleId = o.optString("cid"),
                    actorHex = o.optString("actor"),
                    actorShort = o.optString("short"),
                    targetId = o.optString("target").takeIf { it.isNotEmpty() },
                    snippet = o.optString("snippet"),
                    createdAt = o.optLong("at"),
                    emoji = o.optString("emoji").takeIf { it.isNotEmpty() },
                )
            }
        }.getOrDefault(emptyList())
    }

    private fun saveAppRows(rows: List<Row>) {
        val arr = JSONArray()
        for (r in rows) {
            arr.put(JSONObject()
                .put("id", r.id).put("kind", r.kind).put("cid", r.circleId)
                .put("actor", r.actorHex).put("short", r.actorShort)
                .put("target", r.targetId ?: "").put("snippet", r.snippet)
                .put("at", r.createdAt).put("emoji", r.emoji ?: ""))
        }
        runCatching { prefs.edit().putString(KEY_APP_ROWS, arr.toString()).apply() }
    }

    // ---- Engine rows ---------------------------------------------------------------------------

    /**
     * Something was ingested — recompute the merged list off-main, debounced so an ingest burst
     * costs one reduce, not one per envelope. `social.activity()` copies each circle's events out
     * under the lock and reduces with it released, so this never stalls the engine.
     */
    fun poke(social: HavenSocial?) {
        if (!ready || refreshPending) return
        refreshPending = true
        scope.launch {
            kotlinx.coroutines.delay(1_500)
            refreshPending = false
            refresh(social)
        }
    }

    private fun refresh(social: HavenSocial?) {
        val now = System.currentTimeMillis()
        val engineRows: List<Row> = if (social != null) {
            runCatching {
                social.activity(((now - ENGINE_WINDOW_MS).coerceAtLeast(0L)).toULong(), now.toULong())
            }.getOrDefault(emptyList()).map {
                Row(
                    id = it.id, kind = it.kind, circleId = it.circleId,
                    actorHex = it.actorHex, actorShort = it.actorShort,
                    targetId = it.targetId, snippet = it.snippet,
                    createdAt = it.createdAt.toLong(), emoji = it.emoji,
                )
            }
        } else rows.filter { it.kind !in setOf("connect", "circle") }   // keep last engine merge
        val merged = (engineRows + loadAppRows())
            .sortedWith(compareByDescending<Row> { it.createdAt }.thenBy { it.id })
            .take(500)
        if (merged != rows) {
            rows = merged
            bump()
        }
    }

    private fun bump() {
        scope.launch(Dispatchers.Main) { version.intValue++ }
    }
}
